Simplified/AxisService/NYPLAxisContentDownloader.swift

```diff
@@ -0,0 +1,153 @@
+//
+//  NYPLAxisContentDownloader.swift
+//  Simplified
+//
+//  Created by Raman Singh on 2021-04-22.
+//  Copyright © 2021 NYPL Labs. All rights reserved.
+//
+
+import Foundation
+
+protocol NYPLAxisContentDownloading {
+  func downloadItem(from url: URL,
+                    _ completion: @escaping (Result<Data, Error>) -> Void)
+}
+
+
+/// A content downloader which stops downloading content upon receiving a failure.
+class NYPLAxisContentDownloader: NYPLAxisContentDownloading {
+  
+  private let tasks: ThreadSafeDictionaryContainer<URL, URLSessionDataTask>
+  private let handlers: ThreadSafeDictionaryContainer<URL, (Result<Data, Error>) -> Void>
+  private let numberOfAttempts: ThreadSafeDictionaryContainer<URL, Int>
+  private let errored: ThreadSafeValueContainer<Bool>
+  let networkExecutor: NYPLAxisNetworkExecuting
+  
+  init(networkExecuting: NYPLAxisNetworkExecuting) {
+    self.tasks = ThreadSafeDictionaryContainer()
+    self.handlers = ThreadSafeDictionaryContainer()
+    self.numberOfAttempts = ThreadSafeDictionaryContainer()
+    self.errored = ThreadSafeValueContainer(value: false)
+    self.networkExecutor = networkExecuting
+  }
+  
+  func downloadItem(from url: URL,
+                    _ completion: @escaping (Result<Data, Error>) -> Void) {
+    
+    // We dont want to download anything if it's already being downloaded or if
+    // we experienced an error downloading something before.
+    guard !errored.value, tasks[url] == nil, handlers[url] == nil else {
+      return
+    }
+    
+    numberOfAttempts[url] = (self.numberOfAttempts[url] ?? 0) + 1
+    handlers[url] = completion
+    tasks[url] = self.getTask(for: url)
+  }
+  
+  private func getTask(for url: URL) -> URLSessionDataTask {
+    
+    let request = URLRequest(url: url,
+                             cachePolicy: .reloadIgnoringLocalCacheData,
+                             timeoutInterval: networkExecutor.requestTimeout)
+    
+    let task = networkExecutor.GET(request) { [weak self] (result) in
+      guard let self = self else { return }
+      switch result {
+      case .success(let data):
+        self.handlers[url]?(.success(data))
+        self.tasks[url] = nil
+        self.handlers[url] = nil
+      case .failure(let error):
+        // In poor network conditions, the download might fail on first or
+        // second attempt. We retry 2 more times to make sure we don't
+        // prematurely stop the download process.
+        if
+          let attempts = self.numberOfAttempts[url],
+          attempts < 3,
+          let handler = self.handlers[url] {
+          print("NYPLAxisContentDownloader downloading item from \(url.absoluteString) after \(attempts) failed attempts")
```

Comment:
Should we use `Log` instead of `print` here?

Reply:
We're already logging download failure in `AxisService::downloadItem:from at` method. This print statement is more or less redundant and was placed to make it easy to spot book download failure while debugging.